torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -284,6 +296,12 @@ def _pre_forward(
     # the `grad_fn` is mutated.
     _register_post_backward_hooks(state, handles)
 
+    # Recursively convert args and kwargs to specified precision.
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
+    args, kwargs = _prepare_forward_inputs(
```

Comment:
The main assumption on the performance side from this PR is that the overhead from `_prepare_forward_inputs()` / `_cast_fp_inputs_to_dtype()` is not a concern.

Has anyone confirmed if `t.to(t.dtype)` is fully a no-op with effectively 0 overhead? If not, I wonder if we should change `_cast_fp_inputs_to_dtype()` to guard the `.to()` with an `if` statement:
```
def cast_fn(x: torch.Tensor) -> torch.Tensor:
    def cast_fn(x: torch.Tensor) -> torch.Tensor:
    if not torch.is_floating_point(x) or x.dtype == dtype:
        return x
    y = x.to(dtype)
    # Explicitly copy over `requires_grad` since this runs inside
    # `torch.no_grad()`
    if x.is_leaf:
        y.requires_grad = x.requires_grad
    return y
```
In other words, we do not perform any of the logic if the tensor's dtype already matches. I think with this change, I would be more confident that there will be effectively no performance regression from this. The only additional overhead is a CPU-side recursion over the input data structure via `_apply_to_tensors()` incurred for every FSDP instance.

---

On a related note, I think we need to be careful about using a side stream to do this downcasting. Today, we do not use such a stream, but we have mentioned considering to do so. There may be some strange memory freeing behavior and stream complications. We are investigating this for the post-backward, so I think we should hold off on any stream changes in the near term until we get a better understanding.

Reply:
The implementation of `Tensor.to()` is: 

https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/aten/src/ATen/core/Tensor.cpp#L41-L51

Then I traced where it calls to.

```c++
struct TORCH_API to_dtype {
  using schema = at::Tensor (const at::Tensor &, at::ScalarType, bool, bool, c10::optional<at::MemoryFormat>);
  using ptr_schema = schema*;
  // See Note [static constexpr char* members for windows NVCC]
  STATIC_CONSTEXPR_STR_INL_EXCEPT_WIN_CUDA(name, "aten::to")
  STATIC_CONSTEXPR_STR_INL_EXCEPT_WIN_CUDA(overload_name, "dtype")
  STATIC_CONSTEXPR_STR_INL_EXCEPT_WIN_CUDA(schema_str, "to.dtype(Tensor(a) self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)")
  static at::Tensor call(const at::Tensor & self, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format);
  static at::Tensor redispatch(c10::DispatchKeySet dispatchKeySet, const at::Tensor & self, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format);
};
```

```c++
// aten::to.dtype(Tensor(a) self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)
at::Tensor to_dtype::redispatch(c10::DispatchKeySet dispatchKeySet, const at::Tensor & self, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format) {

    static auto op = create_to_dtype_typed_handle();
    return op.redispatch(dispatchKeySet, self, dtype, non_blocking, copy, memory_format);
}

```

```c++
// aten::to.dtype(Tensor(a) self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)
static C10_NOINLINE c10::TypedOperatorHandle<to_dtype::schema> create_to_dtype_typed_handle() {
  return c10::Dispatcher::singleton()
      .findSchemaOrThrow(to_dtype::name, to_dtype::overload_name)
      .typed<to_dtype::schema>();
}
```

https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/aten/src/ATen/core/dispatch/Dispatcher.h#L489-L492

https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/aten/src/ATen/core/dispatch/Dispatcher.h#L641-L656


At this point, I didn't continue. Even if it does not create any CUDA kernel, the above is sufficient complexity + python/cpp context switch. Let's avoid that by using your suggestion above.

Reply:
Agreed, looks like even if tensor.to() is a noop, we can avoid a python/C++ switch and GIL release / reacquire here.